src/arcade/core/util/Plane.java

```diff
@@ -0,0 +1,82 @@
+package arcade.core.util;
+
+import sim.util.Int3D;
+
+/**
+ * A plane in 3D space.
+ */
+
+public final class Plane {
+    /** A point on the plane. */
+    public final Int3D point;
+
+    /** The normal vector to the plane. */
+    public final Int3D normalVector;
+
+    /**
+     * Creates a plane from a point and a vector.
+     *
+     * @param point  a point on the plane
+     * @param normalVector  the normal vector to the plane
+     */
+    public Plane(Int3D point, Int3D normalVector) {
+        this.point = point;
+        this.normalVector = normalVector;
+    }
+
+    /**
+     * Determines whether a point is on the plane.
+     *
+     * @param p  the point to test
+     * @return  {@code true} if the point is on the plane
+     *        {@code false} otherwise
+     */
+    public boolean isOnPlane(Int3D p) {
+        return (p.getX() - point.getX()) * normalVector.getX()
+                + (p.getY() - point.getY()) * normalVector.getY()
+                + (p.getZ() - point.getZ()) * normalVector.getZ() == 0;
+    }
+
+    /**
+     * Determines distance from a point to the plane.
+     *
+     * @param p  the point
+     * @return  the distance from the point to the plane.
+     *          The distance is positive if the point is on
+     *          the same side of the plane as the normal vector
+     *         and negative if it is on the opposite side.
+     */
+    public double distanceToPlane(Int3D p) {
+        return (p.getX() - point.getX()) * normalVector.getX()
+                + (p.getY() - point.getY()) * normalVector.getY()
+                + (p.getZ() - point.getZ()) * normalVector.getZ();
+    }
+
+    /**
+     * Determines if two planes are equal.
+     *
+     * @param obj  the plane to compare
+     * @return {@code true} if the planes are equal, {@code false} otherwise
+     */
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || getClass() != obj.getClass()) {
+            return false;
+        }
+        Plane other = (Plane) obj;
+        return point.equals(other.point) && normalVector.equals(other.normalVector);
+    }
+
+    /**
+     * Returns a hash code for the plane.
+     *
+     * @return  a hash code for the plane
+     */
+    @Override
+    public int hashCode() {
```

Comment:
I don't know that you need to hash on `Plane` objects

Reply:
I only added this because I got an error when I implemented the equals without implementing the hash. Can I safely ignore this error?

Reply:
Right, forgot that if you have to override `hashCode` if you override `equals`. I would revisit this method to make it more clear how the hash is calculated, maybe:

```java
public int hashCode() {
      return referencePoint.x + (referencePoint.y << 5) + (referencePoint.z << 10)
              + (normalVector.x + 15) + (normalVector.y << 20) + (normalVector.z << 23);
  }
```